### ISSUE TITLE
2.4.11 and browser default focus indicators

### ISSUE THREAD

**Issue by JamesCatt** [w3c/wcag#1854]
I'm not convinced that this requirement, by effectively banning the use of browser default focus indicators, is actually going to lead to a better outcome for users.

A few years ago, when all the browsers had terrible focus indicators, it made a lot of sense—but things have changed. Chrome and Edge now have an _excellent_ focus indicator. In fact, I would argue that it's superior to anything that a developer can achieve with CSS. And it's _almost certainly far superior to what the vast majority of site authors would implement when complying with this requirement_.

Consider the following benefits of the new focus indicator in Chrome/Edge:

- two-tone, white/dark, so it's highly visible on any color background
- doesn't rely on color vision to be visible
- automatically adapts to user-defined color schemes such as with Windows High Contrast Mode
- is consistent across all interface components

Now in theory, yes it is possible to create a custom focus indicator that ticks all of those boxes with CSS. But doing this requires a decent amount of effort and know-how. Realistically, not many sites will do this owing to lack of resources, knowledge, or just general apathy. The end result to the user is a worse, less-accessible experience than if the site just continued using the browser default.

As an example, just try tabbing through the items at the top of the page here on Github. Most of the top links use the default indicator and the focus is highly visible as a result. But then when you get to the controls for "Watch", "Star", "Code", "Issues", etc, this disappears and you get a variety of custom indicators which are far less visible and completely lack consistency. Yes I know that these indicators would not comply with 2.4.11, but even if they did they would still likely be worse than just using the Chrome/Edge default (and still lacking in consistency).

I realize this still leaves a gap where browsers such as Firefox or Safari are concerned. But wouldn't the better course of action to be to work with Mozilla/Apple to get them to improve their defaults? Or better yet, get them to add the option for users to easily customize the indicator?

To be clear, I think this requirement as drafted is excellent for scenarios where a site author _does_ decide to use a custom focus indicator. But when we consider the imperative to put the user first, I worry that the end result of effectively banning browser defaults will potentially be worse for accessibility. This requirement effectively forces authors to implement a custom indicator that _at most_ can only ever be as good as the Chrome/Edge default, and in most cases will be worse.

This feels like we're trying to cure the headache by cutting off the head.

**Comment by patrickhlauke** [comment]
> I'm not convinced that this requirement, by effectively banning the use of browser default focus indicators, is actually going to lead to a better outcome for users.

where is 2.4.11 normatively banning default focus indicators?

the closest it comes is the mention in the intent of the understanding doc https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance-minimum.html but that's far from a ban

> The default focus indicators in some browsers can be difficult to see, such as a 1 pixel dotted outline, or a blue indicator which happens to be on a blue background.It is generally best to either define a keyboard focus style which meets this criterion, or test the focus styles across the browsers that are relied upon for conformance.

**Comment by mraccess77** [comment]
2.4.11 doesn't ban default - but there is not an exception (at this time and to my knowledge) for them if they do not conform.

**Comment by JAWS-test** [comment]
There are still enough cases where Chrome's default focus is not visible at all, e.g. 

```
<a href=# style="outline:2px solid black;outline-offset:2px;">link</a>
<button style="border:2px solid black;">button</button>
```

**Comment by patrickhlauke** [comment]
> If you don't implement a custom indicator, wouldn't your site would fail when used with Firefox or Safari, as their default indicators fail the guideline?

so there we go...because in those browsers the indicators are notoriously rubbish. so while it's true that "things have changed", you've basically just proven the point that they haven't universally, across the board...

**Comment by patrickhlauke** [comment]
> In the current scenario, you'd have to override the default one in Chrome in favor of a cross-browser solution. That doesn't make a whole lot of sense when the chrome one isn't achievable in CSS.

so the alternative is...what? just accept that on browsers with rubbish default focus indication "it is what it is"?

**Comment by PAFaieta** [comment]
Kind of. I mean, there are specific tags you can use, but thinking to the future, it doesn't make sense to develop around the issues of the rest of the field and hope.

That's part of what James was getting at because we either focus our efforts on developing a holistic style ourselves under the assumption that the field won't catch up, or accept it and then open issues with the other browsers to say that their solution needs improvement. Knowing that, I would agree with his solution to work with the third parties to get something better natively built.

**Comment by patrickhlauke** [comment]
once all browsers have their own good focus indicators, this SC will be satisfied without the author needing to do anything...so this *is* thinking to the future, no?

**Comment by patrickhlauke** [comment]
and noting: the "it's not possible for authors to define styles that are as good as the good defaults in Chrome etc" may be true, but only if we're talking about creating a single focus style that adapts well to both light and dark backgrounds. a one-size-fits-all single style. authors can, however, define explicit styles for the different situations they have on their site specifically, since they're in full control...

**Comment by JamesCatt** [comment]
> 2.4.11 doesn't ban default focus indicators- but there is not an exception (at this time and to my knowledge) for them if they do not conform.

@mraccess77 ...and by extension that means you have to override them. The SC may not explicitly ban them, but for all practical purposes authors are forced to replace them.

> I consider it progress that the browser's default focus is no longer automatically sufficient when it is poorly visible. On the other hand, I find it very legitimate to ask how to evaluate the default focus when it meets 2.4.11 in some browsers and not in others. Does it then always need to use its own focus indicator to make it work across browsers? Or must different CSS be used depending on the browser?

@JAWS-test Completely agree re:progress, but like I said if it results in everyone nuking the Chrome/Edge indicator then it's only progress for some (i.e., users of other browsers), and a regression for others. The last point kind gets to the crux of the issue—implementing different focus styling for some browsers and not others would require the use of hacks and/or non-standard CSS that shouldn't be encouraged. Realistically most site owners aren't going to want a different set of styles for different browsers anyway.

> There are still enough cases where Chrome's default focus is not visible at all, e.g.
> 
> ```
> <a href=# style="outline:2px solid black;outline-offset:2px;">link</a>
> <button style="border:2px solid black;">button</button>
> ```

@JAWS-test I don't think the first example is a valid scenario. As soon as you set `outline` you're modifying the default indicator by definition. The `border` is a valid consideration though.

> so there we go...because in those browsers the indicators are notoriously rubbish. so while it's true that "things have changed", you've basically just proven the point that they haven't universally, across the board...

@patrickhlauke As with my previous point, the likely end result here would be an improvement for Firefox/Safari users, but a regression for Chrome/Edge users. Like I said, why not work with the vendors to fix the problem at the source (i.e,. the user agent, where a solution would be far more robust anyway) instead of putting the onus on site authors and potentially making things worse for some users?



> and noting: the "it's not possible for authors to define styles that are as good as the good defaults in Chrome etc" may be true, but only if we're talking about creating a single focus style that adapts well to both light and dark backgrounds. a one-size-fits-all single style. authors can, however, define explicit styles for the different situations they have on their site specifically, since they're in full control...

@patrickhlauke What about high contrast mode? Not to mention lack of consistency—the Chrome indicator is easily identifiable because it's always the same. Consider the example I cited here on Github—the custom styling they have is radically different (and frankly hard to follow IMO) depending on the component. As I said, it's certainly possible for authors to implement something that will work everywhere, as well as the Chrome/Edge one, but realistically how many actually will?

**Comment by patrickhlauke** [comment]
>  Like I said, why not work with the vendors to fix the problem at the source (i.e,. the user agent, where a solution would be far more robust anyway) instead of putting the onus on site authors and potentially making things worse for some users?

who would be doing this work? it's not like the whole community hasn't pushed for changes before. and noting that browser vendors themselves are part of the wider W3C/WAI effort. and despite this, we still are where we are.

**Comment by JamesCatt** [comment]
> who would be doing this work? it's not like the whole community hasn't pushed for changes before. and noting that browser vendors themselves are part of the wider W3C/WAI effort. and despite this, we still are where we are.

@patrickhlauke Fair point.

In any case, I just felt this should be called out because what it comes down to for me is this:

- This SC, as written, will likely have the unintended side effect of making a **lot** of websites **less** accessible for a **lot** of people.
- Furthermore, it will likely lock in this less-than-ideal state of affairs and stymie further progress because authors will feel compelled to implement their own indicators to ensure compliance in all browsers (i.e., even if Firefox implements a better default tomorrow, everyone will _still_ have to override it to make sure they're compliant in Safari, or Samsung Internet, or Opera, etc)
- At the end of the day, if the working group thinks this is an acceptable trade-off, then fair enough. But I personally think it's the wrong way to go. I feel there has to be a better way to attack this problem, even if I'm not smart enough to figure it out.
- Whatever the ultimate decision is, I think it should be explicitly addressed in the SC and not be left open to interpretation. Otherwise, as @JAWS-test pointed out, it could be unclear to auditors how to evaluate a site that passes in one browser but fails in another.

**Comment by mraccess77** [comment]
"other browsers to say that their solution needs improvement. Knowing that, I would agree with" We've had open issues for years with Mozilla.

WCAG conformance requirements do not require content conform in all browsers that are out there - only one combination is required to make a conformance claim.  If we were required to conform to all sorts of browsers then most content would never pass for a multitude of reasons.   

I also challenge the notion that it's not possible to create a two color indicator that that generally works as well as Chrome's does.

**Comment by JAWS-test** [comment]
@mraccess77 

> WCAG conformance requirements do not require content conform in all browsers that are out there - only one combination is required to make a conformance claim

* How many and which browsers should be supported to be WCAG compliant?
* Ccombination of what? UA and AT?

I think that not only one browser must be supported:

> Content posted to the public Web may need to work with a broader range of user agents and assistive technologies, including older versions. (https://www.w3.org/TR/UNDERSTANDING-WCAG20/conformance#uc-accessibility-support-head)

**Comment by JAWS-test** [comment]
I notice 2 more problems with Chrome's default focus indicator:
* the focus frame does not scale when the page is zoomed, i.e. at 200% zoom the focus frame stays thin as it is
* the focus frame has no spacing, i.e. overlaps the text of the focused element at the edge

**Comment by alastc** [comment]
Hi @JamesCatt,

Edge (and then Chrome) changed their indicators shortly after we started working on 2.4.11, we discussed this aspect then, and decided it was still worth pursuing. 

The reasoning has mostly been covered in the thread, but to summarise:

- Even just using Edge/Chrome, there are still instances where the focus indicator can be virtually invisible.
- Other browsers (particularly Firefox) have notoriously difficult to see focus indicators, and there is no sign of that changing despite bugs being raised about it for many years.
- If they do change, the requirement will be fulfilled and authors won't need to worry about it.

I strongly disagree that this requirement will make sites less accessible. Since this SC was proposed we've been including this as a 'future requirement' in our audits, and it's been easy to catch (because the indicators are generally rubbish at the moment) and the resolutions are almost always straightforward [1].

-Alastair

1] There have been some interesting edge cases such as on Gov.uk where they used a thick underline on a very small & narrow link, but we managed to resolve that.

**Comment by JamesCatt** [comment]
@alastc fair enough, and thanks for the background. I'm glad the question was actively considered already.

Perhaps I've overstated the case, but I guess it just rubs me the wrong way that now that we're finally starting to see progress amongst browsers on this point we're effectively going to nuke it. And I do worry that this may reduce the incentive for browsers to bother making improvements to their indicators going forward (meaning that users will continue to suffer on sites that don't bother to comply with WCAG). That's not really a reason to not do it though.

It does still leave the hypothetical question of what the threshold is for satisfying the requirement by relying on the default in the event that some browsers improve their indicators.

**Comment by mbgower** [comment]
To Alastair's summary, I'll just add the point that there are still governments standardized on older browsers. If we allow exceptions for user agent defaults, we are sentencing employees of these organizations to further years of crap user experiences.

In a situation where an organization _has_ standardized their intranet on a browser with better default indicators, then with this SC their design teams need only concentrate on resolving the gaps in the browser's implementation until such time as those are fixed.

**Comment by alastc** [comment]
Hi everyone, the response above was approved by the group, so closing this issue.

**Comment by escetic** [comment]
Hi @alastc, can you clarify which group approved the response.  If this isn't the official way to contribute or help shape the guidelines then what is the process, can you point to how to join this off-thread committee which approves/disapproves things?  Thank you.
